shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -619,6 +630,21 @@ - (UITextRange*)rangeEnclosingPosition:(UITextPosition*)position
 }
 
 - (UITextRange*)lineEnclosingPosition:(UITextPosition*)position {
+  // TODO(hellohuanlin): Remove iOS 17 check. The same logic should apply to older versions too.
+  if (@available(iOS 17.0, *)) {
+    // The end of document (with backward affinity) should not be part of any line, since end is
+    // exclusive. This is to fix 2 bugs:
+    // 1. Keyboard freezes when switching languages:
+    // https://github.com/flutter/flutter/issues/134716
+    // 2. Auto correction candidate menu does not show up in iOS 17:
+    // https://github.com/flutter/flutter/issues/132594
+    // The end of document with forward affinity should still return the last line, which is used
```

Comment:
What's `endOfDocument` supposed to return? If my text is `abcd` then with the current implementation `endOfDocument` is `(4, backward)`, so `beginningOfDocument` and `endOfDocument` form a closed interval that describes valid cursor positions within the document?

If that's the case shouldn't we return `nil` when `position` is greater than `(4, backward)`, and still return a non-nil value when it's `(4, backward)`?

Reply:
According to this comment, 
https://github.com/flutter/engine/blob/b28032c157ee7c3ea06a297e967b5d892fcfc88d/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm#L1892-L1894

`backward` translates to `TextAffinity.upstream`. So `beginningOfDocument` and `endOfDocument` is a close interval in the current implementation. Returning `nil` for `endOfDocument` doesn't seem right (unless `endOfDocument` should be made exclusive).